pennylane/ops/functions/matrix.py

```diff
@@ -134,13 +134,27 @@ def _matrix(tape, wire_order=None):
     interface = qml.math._multi_dispatch(params)
 
     wire_order = wire_order or tape.wires
-    n_wires = len(wire_order)
 
     # initialize the unitary matrix
-    unitary_matrix = qml.math.eye(2**n_wires, like=interface)
+    unitary_matrix = qml.math.eye(2 ** len(wire_order), like=interface)
 
+    broadcasted_unitary = False
     for op in tape.operations:
         U = matrix(op, wire_order=wire_order)
-        unitary_matrix = qml.math.tensordot(U, unitary_matrix, axes=[[-1], [-2]])
+        broadcasted_U = qml.math.ndim(U) == 3
+        if broadcasted_U and broadcasted_unitary:
+            # If both, U and unitary_matrix are broadcasted, we need a special syntax
+            unitary_matrix = qml.math.stack(
+                [qml.math.dot(u, _unitary) for u, _unitary in zip(U, unitary_matrix)]

```

Comment:
Could this be expressed using a `tensordot`? It seems like we're iterating over an axis and contracting another.

Reply:
I tried and failed :D 
The perspectives I found are:
1. We want to perform a standard matrix `dot`, which works with extra dimensions out of the box (e.g. `(3, 4, 4) @ (3, 4, 4)->(3,4,4)` exactly how we want), but it fails in cases where the broadcasting dimension is the same as the Hilbert space dimension (e.g. `(4, 4, 4) @ (4, 4, 4) -> (4, 4, 4, 4)` which is not what we want).
2. `einsum("...ij,..jk->...ik", A, B)` does exactly what we want, but it's terribly slow.
3. The approach I chose here. We kind of want to "multiply the matrices along the latter pair of axes while taking the diagonal w.r.t to the leading dimensions of both objects", if that makes any sense.

I'll check again whether there is magic in the `axes` kwarg of `tensordot` that I don't know yet.

Reply:
Makes sense! No worries if it doesn't come around, maybe some of these findings could even be included as a comment (e.g., that `einsum` would be slower).

Reply:
Aha! I just found out that `np.matmul` and `np.dot` have different broadcasting rules. `matmul` should be what we want :)